backend/src/main/java/org/eclipse/tractusx/puris/backend/common/edc/logic/service/EdcAdapterService.java

```diff
@@ -487,6 +541,420 @@ public Response getProxyPullRequest(String url, String authKey, String authCode,
+                        }
+                    } catch (Exception e) {
+                        log.error("Error in ItemStock Submodel 2.0.0 transfer request for partner " + partner.getBpnl(), e);
+                    }
+
+
+                } catch (Exception e) {
+                    log.error("Error in ItemStock Submodel 2.0.0 transfer request for partner " + partner.getBpnl(), e);
+                }
+            }
+        } finally {
+            if (failed && contractMapping != null) {
+                // invalidate stored data
+                contractMapping.setItemStockAssetId(null);
+                contractMapping.setItemStockContractId(null);
+                contractMapping.setItemStockEdcProtocolUrl(null);
+                contractMapping.getMaterialToHrefMapping().put(materialNumber, null);
+                edcContractMappingService.update(contractMapping);
+                log.warn("ItemStock Submodel request for material " + material.getOwnMaterialNumber() + " at partner " +
+                    partner.getBpnl() + " failed. Invalidating contract data. You may want to retry this request");
+            }
+        }
+        return null;
+    }
+
+    private boolean negotiateForPartnerItemStockSubmodel(MaterialPartnerRelation mpr, DirectionCharacteristic direction) {
+        Partner partner = mpr.getPartner();
+        try {
+            EdcContractMapping contractMapping = edcContractMappingService.find(mpr.getPartner().getBpnl());
+            String itemStockEdcUrl = contractMapping.getItemStockEdcProtocolUrl();
+            String itemStockAssetId = contractMapping.getItemStockAssetId();
+            if (itemStockEdcUrl == null || itemStockAssetId == null) {
+                log.info("Need ItemStock Submodel Edc URL for " + mpr.getPartner().getBpnl());
+                switch (direction) {
+                    case INBOUND -> getPartnerAasForProduct(mpr.getPartner(), mpr.getPartnerMaterialNumber());
+                    case OUTBOUND -> getPartnerAasForMaterial(mpr.getPartner(), mpr.getPartnerMaterialNumber());
+                }
+                contractMapping = edcContractMappingService.find(mpr.getPartner().getBpnl());
+                itemStockEdcUrl = contractMapping.getItemStockEdcProtocolUrl();
+                itemStockAssetId = contractMapping.getItemStockAssetId();
+            }
+            var responseNode = getCatalog(itemStockEdcUrl);
+            var catalogArray = responseNode.get("dcat:dataset");
+            // If there is exactly one asset, the catalogContent will be a JSON object.
+            // In all other cases catalogContent will be a JSON array.
+            // For the sake of uniformity we will embed a single object in an array.
+            if (catalogArray.isObject()) {
+                catalogArray = objectMapper.createArrayNode().add(catalogArray);
+            }
+            JsonNode targetCatalogEntry = null;
+            for (var entry : catalogArray) {
```

Comment:
One thing we can keep in mind: We could also already query the catalog with an assetId filter and then check for supported FrameworkAgreements. As the future edc versions (edc 0.7.0) coming soon, should include nested json in the query spec, we can go for a refactoring then. That would also help in looking up the DTR.

Reply:
Yes, that's one of the future improvements we will add in the coming months or weeks. But I think in the short term, our next step is here to clean up a bit, i.e. remove a lot of methods and endpoints that were solely related to the older ItemStock and are now obsolete. :)